src/main/java/htsjdk/samtools/cram/structure/SubstitutionMatrix.java

```diff
@@ -71,52 +64,37 @@
     // The expanded in-memory matrix of substitution codes. In order to enable quick inter-conversion
     // between bases and substitution codes, we use a full square but sparse matrix that covers
     // the entire symbol space in order to allow them to be directly indexed by (refBase, readBase)
-    // pairs. Note that this array can be indexed using upper or lower case bases.
+    // pairs. Note: since this implementation doesn't *generate* substitutions for lower case bases,
+    // this array can only be used to find a substitution code for lower case reference bases.
     private final byte[][] codeByBase = new byte[SYMBOL_SPACE_SIZE][SYMBOL_SPACE_SIZE];
 
     // The expanded in-memory matrix of substitute bases, indexed by (refBase, code) pairs, used when
-    // reading cram records. Note that this array can be indexed using upper or lower case bases.
+    // reading cram records. Note that this array can be indexed using upper or lower case bases to
+    // handle lower case reference substitutions that may have been generated by other CRAM writers.
     private final byte[][] baseByCode = new byte[SYMBOL_SPACE_SIZE][SYMBOL_SPACE_SIZE];
 
     /**
-     * Create a SubstitutionMatrix given a set of base substitution frequencies
-     * @param frequencies array of base substitution frequencies
+     * Create a SubstitutionMatrix given a list of CramCompressionRecord
+     * @param records array of CramCompressionRecord with Substitutions
      */
-    public SubstitutionMatrix(final long[][] frequencies) {
-        // get the substitution code vector for each base
-        for (final byte base : SUBSTITUTION_BASES_UPPER) {
+    public SubstitutionMatrix(final List<CramCompressionRecord> records) {
+        initializeBaseByCode();
+
+        final long[][] frequencies = buildFrequencies(records);
+        for (final SubstitutionBase b : SubstitutionBase.values()) {
             // substitutionCodeVector has a side effect of updating codeByBase
-            encodedMatrixBytes[CODE_ORDER[base]] = substitutionCodeVector(base, frequencies[base]);
+            encodedMatrixBytes[b.ordinal()] = substitutionCodeVector(b.getBase(), frequencies[b.getBase()]);
         }
 
-        for (final byte[] base : baseByCode) {
-            Arrays.fill(base, (byte) 'N');
-        }
-        for (int i = 0; i < SUBSTITUTION_BASES_UPPER.length; i++) {
-            final byte r = SUBSTITUTION_BASES_UPPER[i];
-            for (final byte b : SUBSTITUTION_BASES_UPPER) {
-                if (r == b) {
-                    continue;
+        for (final SubstitutionBase r : SubstitutionBase.values()) {
+            for (final SubstitutionBase b : SubstitutionBase.values()) {
+                if (r != b) {
+                    baseByCode[r.getBase()][codeByBase[r.getBase()][b.getBase()]] = b.getBase();
```

Comment:
getBase() will return a ```byte``` whose value is -128 to 127. Obviously negatives are going to cause a problem here therefore In practice here you only get 0 to 127 however these arrays are initialized to store from 0 to 256. So you can change the array lengths to 128 without further changes and add the "documented" constraint that getBase() won't ever return a negative... or change its return type to ```char``` that is never a negative (is the unsigned short equivalent in Java)...